ci/docker/vmcontrol.py

```diff
@@ -0,0 +1,269 @@
+        @param timeout: in seconds, if None or 0 wait forever
+        @return: True of False, if timeout is None may return only True or
+                 throw unhandled network exception
+    """
+    import socket
+    import errno
+    import time
+    log = logging.getLogger('wait_ssh_open')
+    sleep_s = 0
+    if timeout:
+        from time import time as now
+        # time module is needed to calc timeout shared between two exceptions
+        end = now() + timeout
+
+    while True:
+        log.debug("Sleeping for %s second(s)", sleep_s)
+        time.sleep(sleep_s)
+        s = socket.socket()
+        try:
+            if keep_waiting and not keep_waiting():
+                log.debug("keep_waiting() is set and evaluates to False")
+                return False
+
+            if timeout:
+                next_timeout = end - now()
+                if next_timeout < 0:
+                    log.debug("connect time out")
+                    return False
+                else:
+                    log.debug("connect timeout %d s", next_timeout)
+                    s.settimeout(next_timeout)
+
+            log.info("connect %s:%d", server, port)
+            s.connect((server, port))
+            ret = s.recv(1024).decode()
+            if ret and ret.startswith('SSH'):
+                s.close()
+                log.info("wait_ssh_open: port %s:%s is open and ssh is ready", server, port)
+                return True
+            else:
+                log.debug("Didn't get the SSH banner")
+                s.close()
+
+        except ConnectionError as err:
+            log.debug("ConnectionError %s", err)
+            if sleep_s == 0:
+                sleep_s = 1
+            else:
+                sleep_s *= 2
+
+        except socket.gaierror as err:
+            log.debug("gaierror %s",err)
+            return False
+
+        except socket.timeout as err:
+            # this exception occurs only if timeout is set
+            if timeout:
+                return False
+
+        except TimeoutError as err:
+            # catch timeout exception from underlying network library
+            # this one is different from socket.timeout
+            raise
+
+
+def wait_port_open(server, port, timeout=None):
+    """ Wait for network service to appear
+        @param server: host to connect to (str)
+        @param port: port (int)
+        @param timeout: in seconds, if None or 0 wait forever
+        @return: True of False, if timeout is None may return only True or
+                 throw unhandled network exception
+    """
+    import socket
+    import errno
+    import time
+    sleep_s = 0
+    if timeout:
+        from time import time as now
+        # time module is needed to calc timeout shared between two exceptions
+        end = now() + timeout
+
+    while True:
+        logging.debug("Sleeping for %s second(s)", sleep_s)
+        time.sleep(sleep_s)
+        s = socket.socket()
+        try:
```

Comment:
Maybe extract this timeout logic into a separate function and pass the logic as lambda to avoid duplicating code?

Reply:
it's a good suggestion, but this is a standalone utility function that is only used once here and it works. I don't want to change it further without good reason.

Reply:
I think we clarified this f2f. You are right, it can be a future improvement indeed. I would like to live it like that at this point, as it's diminishing returns and not a critical component. It has been manually tested, and setting an automated unit test would be costly.

Reply:
Fine with me